azure-kusto-data/azure/kusto/data/__init__.py

```diff
@@ -1,4 +1,148 @@
+        if KustoConnectionStringBuilder._Keywords.application_client_id in self:
+            return self[KustoConnectionStringBuilder._Keywords.application_client_id]
+        return None
+
+    @property
+    def application_key(self):
+        if KustoConnectionStringBuilder._Keywords.application_key in self:
+            return self[KustoConnectionStringBuilder._Keywords.application_key]
+        return None
+
+    @property
+    def authority_id(self):
+        if KustoConnectionStringBuilder._Keywords.authority_id in self:
+            return self[KustoConnectionStringBuilder._Keywords.authority_id]
+        return None
+
+    @authority_id.setter
+    def authority_id(self, value):
+        self[KustoConnectionStringBuilder._Keywords.authority_id] = value
+
+    def _is_valid(self):
+        _ensure_value_is_valid(self[KustoConnectionStringBuilder._Keywords.data_source])
+        error_message_template = "specified '{0}' authentication method has incorrect properties set"
+        if KustoConnectionStringBuilder._Keywords.aad_user_id in self:
+            error_message = error_message_template.format(KustoConnectionStringBuilder._Keywords.aad_user_id.value)
+            _ensure_value_is_valid(self[KustoConnectionStringBuilder._Keywords.password])
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.application_client_id, error_message)
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.application_key, error_message)
+        elif KustoConnectionStringBuilder._Keywords.application_client_id in self:
+            error_message = error_message_template.format(KustoConnectionStringBuilder._Keywords.application_client_id.value)
+            _ensure_value_is_valid(self[KustoConnectionStringBuilder._Keywords.application_key])
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.aad_user_id, error_message)
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.password, error_message)
+        else:
+            error_message = error_message_template.format('Device login')
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.application_client_id, error_message)
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.application_key, error_message)
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.aad_user_id, error_message)
+            self._ensure_value_does_not_exists(KustoConnectionStringBuilder._Keywords.password, error_message)
+
+    def _ensure_value_does_not_exists(self, value, error_message):
+        if value in self:
+            raise AssertionError(error_message)
+
+def _ensure_value_is_valid(value):
+    if not isinstance(value, text_type):
+        raise TypeError
+    if not value or not value.strip():
+        raise ValueError
+
+class KustoClientFactory(object):
```

Comment:
What is the purpose of this class if it only has a single method that is static?

Reply:
I removed it.